### PR TITLE
Consider index signatures as optional properties in contextual type discrimination

### LINKED ISSUE #54498: String constant fails to infer type inside nested unions
# Bug Report

### 🔎 Search Terms

infer contextual string union

### 🕗 Version & Regression Information

- This changed between versions 5.1.3 and 5.0.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.3#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgVSQQiQEEAbGUuAbwCg44YBPMYAfgC44ByVEGDwDcdAL506oSLEQo0WXASIkKMAEK0GTVh248EAW0wBzYDzgAfXgDcEAE2ARzVnnYjYArgeAphYiVLQ8CxscABqmOQewITESHAAvEpxqtRWsSqUaiKS4EHaoQCymGAA8lAAysjG5MAwJIlaVvSMrXAA2gDWwMzcAM4wUNUAutwRUTHKSCJt4oxW49EZ0xLYJANwAO4IMAAWEB5UfQDC6zDcxWWV1bX18UktcACM3I+MIcB6-IJwAPQAVJg+nA1kgBv9fgAaLSiaGiHKgjbbPakE5nC4lcpVJA1OoNB5aF6aNofL6gH5AkFnaGMWFiERAA)

### 💻 Code

```ts repro
export interface UnionAltA {
  type?: 'text';
}

export interface UnionAltB {
  type?: 'image' | 'video' | 'document';
}

export type ValueUnion = UnionAltA | UnionAltB;

export type MapOrSingleton =
  | {
      [key: string]: ValueUnion;
    }
  | ValueUnion;

const withoutAsConst: MapOrSingleton = {
  1: {
    type: 'text' /*as const*/,
  },
};

const withAsConst: MapOrSingleton = {
  1: {
    type: 'text' as const,
  },
};
```

### 🙁 Actual behavior

The declaration of `withoutAsConst` fails, as `type: "text"` fails to be inferred as `"text"`, and instead infers as `string` in this situation:

```
Type 'string' is not assignable to type '"text" | "image" | "video" | "document" | undefined'.
```

### 🙂 Expected behavior

The declaration of `withoutAsConst` should succeed.  Typescript understands that the only valid values in this position are `"text" | "image" | "video" | "document" | undefined` from the contextual type, so the literal `"text"` should match that.

### PR DESCRIPTION
**Pull request by gabritto**
Fixes #54498.
The fix is suggested and described by @nebkat in this comment: https://github.com/microsoft/TypeScript/issues/54498#issuecomment-1575353117.

When we try to discriminate a contextual type, we consider the omission of a property to be a factor to discriminate in `discriminateTypeByDiscriminableItems`'s implementation, but then during that process, when looking up a property in one of the candidate types, we would not consider matching index signatures and therefore we would discard the type with the index signature. Now, when looking up a property in one of the candidate types, we also consider matching on index signatures.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22719,6 +22719,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
         const filtered = contains(include, Ternary.False) ? getUnionType(types.filter((_, i) => include[i])) : target;
         return filtered.flags & TypeFlags.Never ? target : filtered;
+
+        function getTypeOfPropertyOrOptionalIndexSignature(type: Type, name: __String): Type | undefined {
```
Comment (gabritto) [inline src/compiler/checker.ts:22723]: I feel like this function might already exist, but I couldn't find it.

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot test top100
@typescript-bot perf test this

**Review by jakebailey — APPROVED** [review]

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -10161,8 +10161,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return prop ? getTypeOfSymbol(prop) : undefined;
     }
 
-    function getTypeOfPropertyOrIndexSignature(type: Type, name: __String): Type {
-        return getTypeOfPropertyOfType(type, name) || getApplicableIndexInfoForName(type, name)?.type || unknownType;
+    function getTypeOfPropertyOrIndexSignature(type: Type, name: __String, addOptionalityToIndex = false): Type {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:10164]: Since there's only 2 call sites, I would just not make the last parameter optional.

**Comment by ImRodry** [comment]
This didn't land in 5.1.6 right? When is it expected to be released?

**Comment by jakebailey** [comment]
This is in the 5.2 beta (also noted in the issue's milestone).

**Pull request merged**